### PR TITLE
Added nullable type support to the Java mapping

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR adds supports for nullable and non-null types in Java through annotations provided by Checker Framework. These have replaced the old java.util.Optional wrappers we used to use for this. It is part of #611.

It also adds a type-checker to the build system that enforces nullness during compile time. Right now the checker is disabled since there's alot of warnings (which the build system treats as errors and fails) with the mapping, that I'm still in the process of fixing. After fixing them, I'll re-enable the checker.

This PR adds a dependency on the Checker Framework library.

### CODE REVIEW

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 1]
```diff
@@ -962,17 +964,17 @@ public byte readByte()
      * Extracts an optional byte value from the stream.
      *
      * @param tag The numeric tag associated with the value.
-     * @return The optional value (if any).
+     * @return The extracted byte value, or null if it wasn't sent.
      **/
-    public java.util.Optional<Byte> readByte(int tag)
+    public @Nullable Byte readByte(int tag)
     {
         if(readOptional(tag, OptionalFormat.F1))
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java]: Should we switch to the more compact ` return readOptional(...) ? readByte() : null;` ?

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 2]
```diff
@@ -962,17 +964,17 @@ public byte readByte()
      * Extracts an optional byte value from the stream.
      *
      * @param tag The numeric tag associated with the value.
-     * @return The optional value (if any).
+     * @return The extracted byte value, or null if it wasn't sent.
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java]: I don't like `sent` here: it's null when there is no entry for the provided tag in the tagged members map. 
Maybe the simplest is `The extracted Byte`, and it's obvious from the API it's nullable.

In the future, we'll probably have 3 overloads:
 - byte readByte() // non-nullable byte   (supported by all Ice encodings)
 - @Nullable Byte readNullableByte() // nullable Byte (supported only by Ice encoding >= 2.0)
 - @Nullable Byte readNullableByte(int tag) // nullable Byte (supported only by Ice encoding >= 1.1)

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 3]
```diff
@@ -1239,18 +1241,18 @@ public short readShort()
      * Extracts an optional short sequence from the stream.
      *
      * @param tag The numeric tag associated with the value.
-     * @return The optional value (if any).
+     * @return The extracted short sequence, or null if it wasn't sent.
      **/
-    public java.util.Optional<short[]> readShortSeq(int tag)
+    public short @Nullable[] readShortSeq(int tag)
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java]: Short missing?

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 4]
```diff
@@ -1164,17 +1166,17 @@ public boolean readBool()
      * Extracts an optional boolean sequence from the stream.
      *
      * @param tag The numeric tag associated with the value.
-     * @return The optional value (if any).
+     * @return The extracted boolean sequence, or null if it wasn't sent.
      **/
-    public java.util.Optional<boolean[]> readBoolSeq(int tag)
+    public boolean @Nullable[] readBoolSeq(int tag)
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java]: boolean missing?

File: java/src/Ice/src/main/java/com/zeroc/Ice/OperationMode.java [hunk 5]
```diff
@@ -15,6 +15,8 @@
 
 package com.zeroc.Ice;
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/OperationMode.java]: This file should be generated code

File: java/src/Ice/src/main/java/com/zeroc/Ice/ProtocolVersion.java [hunk 6]
```diff
@@ -118,33 +120,28 @@ static public ProtocolVersion ice_read(InputStream istr)
...
-    static public void ice_write(OutputStream ostr, int tag, java.util.Optional<ProtocolVersion> v)
-    {
-        if(v != null && v.isPresent())
-        {
-            ice_write(ostr, tag, v.get());
-        }
-    }
-
-    static public void ice_write(OutputStream ostr, int tag, ProtocolVersion v)
+    static public void ice_write(OutputStream ostr, int tag, @Nullable ProtocolVersion v)
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/ProtocolVersion.java]: The marshaling code for non-generated types (that were previously generated) should be removed, not updated.

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/RoutableReference.java [hunk 7]
```diff
@@ -77,10 +79,10 @@
     }
 
     @Override
-    public java.util.OptionalInt
+    public @Nullable Integer
     getTimeout()
     {
-        return _overrideTimeout ? java.util.OptionalInt.of(_timeout) : java.util.OptionalInt.empty();
+        return _overrideTimeout ? _timeout : null;
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/RoutableReference.java]: Could we merge _overrideTimeout and _timeout into a single _timeout Integer?

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/RouterInfo.java [hunk 8]
```diff
@@ -78,7 +78,7 @@
         }
 
         com.zeroc.Ice.Router.GetClientProxyResult r = _router.getClientProxy();
-        return setClientEndpoints(r.returnValue, r.hasRoutingTable.orElse(true));
+        return setClientEndpoints(r.returnValue, (r.hasRoutingTable != null? r.hasRoutingTable : true));
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/RouterInfo.java]: Space missing before ?

File: java/src/Ice/build.gradle [hunk 9]
```diff
@@ -26,6 +26,5 @@ apply from: "$project.ext.topSrcDir/java/gradle/library.gradle"
 javadoc.dependsOn(compileJava)
 
 javadoc {
-    classpath = project.sourceSets.main.output
```
Comment (bernardnormier) [inline java/src/Ice/build.gradle]: Does this disable javadoc?

File: java/build.gradle [hunk 10]
```diff
@@ -51,6 +63,18 @@ subprojects {
         excludeDirs -= file(buildDir)
         buildDir.listFiles({d, f ->f != 'generated-src'} as FilenameFilter).each { excludeDirs += it }
     }
+
+    checkerFramework {
```
Comment (bernardnormier) [inline java/build.gradle]: We should target java11+

File: java/test/src/main/java/test/Ice/optional/Test.ice [hunk 11]
```diff
@@ -346,8 +336,6 @@ interface Initial
     bool supportsCsharpSerializable();
 
     bool supportsCppStringView();
-
-    bool supportsNullOptional();
```
Comment (bernardnormier) [inline java/test/src/main/java/test/Ice/optional/Test.ice]: Probably should keep supportsNullOptional() (returns false) for now, for cross-testing.

File: java/src/Ice/build.gradle (hunk 9 above)
Comment (InsertCreativityHere) [reply java/src/Ice/build.gradle]: No, javadocs still build fine. Before we were just restricting what files could be referenced in the docs. I'm planning on doing a more exhaustive comparison Monday, but from when I compared them, they seemed identical.
But this doesn't affect which classes we generate documentation for, just what those documentations are allowed to reference.

File: java/src/Ice/src/main/java/com/zeroc/Ice/OperationMode.java (hunk 5 above)
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/OperationMode.java]: Alright, I'll remove it, and add the correct metadata to the slice file then. Just wanted to make sure before I started scraping things I shouldn't.

**Pull request merged**